File: test/cgi/cgi_files/executor/body.cpp
#include <stdio.h>
#include <unistd.h>

#include <cstdlib>
#include <cstring>
#include <iostream>
#include <string>

int main(int argc, char *argv[], char *envp[]) {
  printf("Status: 200\r\nContent-Type: text/html\r\n\r\n");
  const char *length = std::getenv("CONTENT_LENGTH");
  if (length == NULL) {
    std::cerr << "CONTENT_LENGTH not found\n";
    return 1;
  }
  size_t len = std::stoll(length);
  char buffer[len];

  if (read(0, buffer, len) == -1) {
    std::cerr << "read " << std::strerror(errno) << "\n";
  }
  printf("%s", buffer);
  return 0;
}

File: test/cgi/cgi_files/executor/body.py
#! /usr/bin/python3

import os
import sys


def printMetaVars():
    print("~~~ ALL ENVIRON ~~~", file=sys.stderr)
    for key, value in os.environ.items():
        print(key, ": ", value, file=sys.stderr)
    print("~~~ ALL ENVIRON printed ~~~", file=sys.stderr)


def main():
    print("Status: 200\r\nContent-Type: text/html\r\n\r\n", end="")
    # printMetaVars()
    kLength = os.environ.get("CONTENT_LENGTH")
    if kLength is None:
        print("CONTENT_LENGTH not found")
        exit(1)
    body = sys.stdin.read(int(kLength))
    print(body, end="")


if __name__ == "__main__":
    main()
